File: tests/CodexiumMagnus.Storage.Tests/DbInitializerTests.cs
using System;
using System.IO;
using Microsoft.Data.Sqlite;
using CodexiumMagnus.Storage;
using Xunit;

namespace CodexiumMagnus.Storage.Tests
{
    public class DbInitializerTests
    {
        [Fact]
        public void EnsureCreated_IsIdempotent_AndCreatesTables()
        {
            // Arrange: make a truly temp path
            var tempDir = Path.Combine(Path.GetTempPath(), "cm-storage-tests-" + Guid.NewGuid().ToString("N"));
            Directory.CreateDirectory(tempDir);
            var dbPath = Path.Combine(tempDir, "app.db");

            // Act: call twice, as app does on startup
            DbInitializer.EnsureCreated(dbPath);
            DbInitializer.EnsureCreated(dbPath);

            // Assert: open and check tables
            using var conn = new SqliteConnection($"Data Source={dbPath}");
            conn.Open();

            bool HasTable(string name)
            {
                using var cmd = conn.CreateCommand();
                cmd.CommandText = "SELECT name FROM sqlite_master WHERE type='table' AND name=@n;";
                cmd.Parameters.AddWithValue("@n", name);
                var result = cmd.ExecuteScalar();
                return result is string;
            }

            Assert.True(HasTable("Cartridges"), "Cartridges table should exist");
            Assert.True(HasTable("BibliographyEntries"), "BibliographyEntries table should exist");
            Assert.True(HasTable("UserSettings"), "UserSettings table should exist");
        }
    }
}

File: tests/codexium-magnus-storage-tests/DbInitializerTests.cpp
#include <QtTest/QtTest>
#include <QSqlDatabase>
#include <QSqlQuery>
#include <QDir>
#include <QTemporaryDir>
#include "DbInitializer.h"

using namespace CodexiumMagnus::Storage;

class DbInitializerTests : public QObject {
    Q_OBJECT

private slots:
    void ensureCreated_IsIdempotent_AndCreatesTables();
};

void DbInitializerTests::ensureCreated_IsIdempotent_AndCreatesTables() {
    // Arrange: create temporary directory
    QTemporaryDir tempDir;
    QVERIFY(tempDir.isValid());
    
    QString dbPath = tempDir.filePath("app.db");

    // Act: call twice, as app does on startup
    DbInitializer::ensureCreated(dbPath);
    DbInitializer::ensureCreated(dbPath);

    // Assert: open and check tables
    QSqlDatabase db = QSqlDatabase::addDatabase("QSQLITE", "test_connection");
    db.setDatabaseName(dbPath);
    QVERIFY(db.open());

    QSqlQuery query(db);
    
    // Check Cartridges table
    query.prepare("SELECT name FROM sqlite_master WHERE type='table' AND name=?");
    query.addBindValue("Cartridges");
    QVERIFY(query.exec());
    QVERIFY(query.next());
    QCOMPARE(query.value(0).toString(), QString("Cartridges"));

    // Check BibliographyEntries table
    query.prepare("SELECT name FROM sqlite_master WHERE type='table' AND name=?");
    query.addBindValue("BibliographyEntries");
    QVERIFY(query.exec());
    QVERIFY(query.next());
    QCOMPARE(query.value(0).toString(), QString("BibliographyEntries"));

    // Check UserSettings table
    query.prepare("SELECT name FROM sqlite_master WHERE type='table' AND name=?");
    query.addBindValue("UserSettings");
    QVERIFY(query.exec());
    QVERIFY(query.next());
    QCOMPARE(query.value(0).toString(), QString("UserSettings"));

    db.close();
    QSqlDatabase::removeDatabase("test_connection");
}

QTEST_MAIN(DbInitializerTests)
#include "DbInitializerTests.moc"
